<launch>
  <!-- These parameters are used for the benchmark evaluation. Documentation can be found in parameter_server.cpp -->
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find rgbdslam)/rgbd_benchmark/log_eval.conf"/>
  <env name="ROSCONSOLE_FORMAT" value="[${severity}] Time:[${time}] Thread:[${thread}]: ${message}"/>
  <arg name="debug" default="false"/>
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/time"/-->
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/gdb -ex run -args"/-->
  <arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/xterm -e gdb -ex run -args"/>
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/valgrind -DELeTEME-tool=cachegrind -DELETEME-cachegrind-out-file=/tmp/cachegrind.out"/-->
  <!--arg if="$(arg debug)" name="launch_prefix" value="/usr/bin/valgrind -DELETEME-leak-check=full "/-->
  <arg unless="$(arg debug)" name="launch_prefix" value=""/>
  <node pkg="rgbdslam" type="rgbdslam" name="rgbdslam" cwd="node" required="true" output="screen"  launch-prefix="$(arg launch_prefix)"> 
  	<param name="config/topic_image_mono"              value="/camera/rgb/image_color"/>
  	<param name="config/topic_image_depth"             value="/camera/depth/image"/>
  	<param name="config/topic_points"                  value=""/>
    <param name="config/wide_topic"                    value=""/>;
    <param name="config/wide_cloud_topic"              value=""/>;
  	<param name="config/fixed_frame_name"              value="/map"/>
    <param name="config/odom_frame_name"               value=""/>
    <param name="config/ground_truth_frame_name"       value="/world"/><!--empty string if no ground truth-->
    <param name="config/base_frame_name"               value="/openni_rgb_optical_frame"/> <!-- /openni_camera for hand-held kinect. For robot, e.g., /base_link -->
    <param name="config/fixed_camera"                  value="true"/> <!--is the kinect fixed with respect to base, or can it be moved-->
  	<param name="config/start_paused"                  value="false"/>
  	<param name="config/store_pointclouds"             value="false"/>
  	<param name="config/subscriber_queue_size"         value="9"/>
  	<!--param name="config/depth_scaling_factor"          value="1.0042723"/-->
  	<!--param name="config/feature_detector_type"         value="SURF"/-->
  	<!--param name="config/feature_extractor_type"        value="SURF"/-->
  	<param name="config/surf_max_iterations"           value="1"/>
    <param name="config/matcher_type"                  value="FLANN"/> <!-- FLANN (not avail for ORB features), SIFTGPU (only for SIFTGPU detector)  or BRUTEFORCE-->
  	<!--param name="config/bagfile_name"                  value="/home/endres/tmp/rgbd_datasets/rgbd_dataset_freiburg1_desk.bag"/-->
    <param name="config/nn_distance_ratio"             value="0.60"/> <!-- Feature correspondence is valid if distance to nearest neighbour is smaller than this parameter times the distance to the 2nd neighbour -->
  	<param name="config/max_keypoints"                 value="600"/>
  	<param name="config/min_keypoints"                 value="400"/>
    <param name="config/sufficient_matches"            value="601"/><!-- Instead of matching all new descriptors against all of a previous node in one step, sufficient_matches+100 of the new descriptors are iteratively compared to all of the previous node until sufficient_matches are found. Setting this parameter low (e.g. 2x min_matches) speeds up comparisons of frames with many matches, but with a potential loss of accuracy, as the transformation is estimated from less features. Set it to max_keypoints for maximum accuracy -->
    <param name="config/min_translation_meter"         value="-1.0"/><!--disabled -->
    <param name="config/min_rotation_degree"           value="-1.0"/><!--disabled -->
  	<param name="config/min_time_reported"             value="0.005"/>

    <param name="config/predecessor_candidates"        value="8"/><!-- search through this many immediate predecessor nodes for corrspondences -->
    <param name="config/neighbor_candidates"           value="8"/><!-- search through this many graph neighbour nodes for corrspondences -->
    <param name="config/min_sampled_candidates"        value="8"/><!-- search through this many uniformly sampled nodes for corrspondences -->
  <param name="config/max_connections"               value="8"/><!-- stop after this many succesfully found spation relations -->
  	
  	<param name="config/drop_async_frames"             value="false"/>
  	<param name="config/min_matches"                   value="10"/>
  	<param name="config/max_dist_for_inliers"          value="1.5"/>
  	<param name="config/ransac_iterations"             value="5000"/>
  	<param name="config/use_gui"                       value="false"/>
  	<param name="config/use_glwidget"                  value="false"/> <!-- incompatible with voxelgrid filter > 0 -->
  	<param name="config/concurrent_node_construction"  value="true"/>
  	<param name="config/concurrent_edge_construction"  value="true"/>
  	<param name="config/concurrent_optimization"       value="false"/>
    <param name="config/optimizer_skip_step"           value="1000000000"/><!-- optimize at end only -->
    <param name="config/backend_solver"                value="auto"/>
  	<param name="config/data_skip_step"                value="1"/>
  	<param name="config/visualization_skip_step"       value="3"/>
    <param name="config/squared_meshing_threshold"     value="0.0081"/><!-- in visualization don't triangulate over distances higher than the root of this parameter -->
    <param name="config/batch_processing"              value="true"/>    <!--store results and close after bagfile has been processed-->
    <param name="config/keep_all_nodes"                value="true"/> 
    <param name="config/use_icp"                       value="false"/> <!-- Ignored if ICP is not compiled in (see top of CMakeLists.txt) -->
    <param name="config/gicp_max_cloud_size"           value="25000"/> <!-- Ignored if ICP is not compiled in (see top of CMakeLists.txt) -->
  </node>
  <!--include file="$(find openni_camera)/launch/kinect_frames.launch"/-->
</launch>
